src/script/env-script.tsx

```diff
@@ -0,0 +1,35 @@
+// XXX: Blocked by https://github.com/vercel/next.js/pull/58129
+// import { headers } from 'next/headers';
+import { type FC } from 'react';
+
+import { type ProcessEnv } from '../typings/process-env';
+
+type EnvScriptProps = {
+  env: ProcessEnv;
+};
+
+/**
+ * Sets the provided environment variables in the browser. If an nonce is
+ * available, it will be set on the script tag.
+ *
+ * Usage:
+ * ```ts
+ * <head>
+ *   <EnvScript env={{ NODE_ENV: 'test', API_URL: 'http://localhost:3000' }} />
+ * </head>
+ * ```
+ */
+export const EnvScript: FC<EnvScriptProps> = ({ env }) => {
+  // It's strongly recommended to set a nonce on your script tags. For
+  // const nonce = headers().get('x-nonce') ?? undefined;
+
+  return (
+    <script
+      data-testid="env-script"
+      // nonce={nonce}
+      dangerouslySetInnerHTML={{
+        __html: `window.__ENV = ${JSON.stringify(env)}`,
```

Comment:
You could also consider making the key to the environment variables a shared constant.

```tsx
import { PUBLIC_ENV_KEY } from 'contants.ts';

// ...
  return (
    <script
      data-testid="env-script"
      // nonce={nonce}
      dangerouslySetInnerHTML={{
        __html: `window[${PUBLIC_ENV_KEY}] = ${JSON.stringify(env)}`,
```